engine/flink/components/base/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/UnionTransformer.scala

```diff
@@ -1,55 +1,44 @@
 package pl.touk.nussknacker.engine.flink.util.transformer
 
-import cats.data.ValidatedNel
+import cats.data.{Validated, ValidatedNel}
 import com.typesafe.scalalogging.LazyLogging
 import org.apache.flink.api.common.functions.FlatMapFunction
 import org.apache.flink.streaming.api.scala._
 import org.apache.flink.streaming.runtime.operators.windowing.TimestampedValue
 import org.apache.flink.util.Collector
 import pl.touk.nussknacker.engine.api._
-import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.NodeId
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.{CannotCreateObjectError, NodeId}
 import pl.touk.nussknacker.engine.api.context.{ContextTransformation, JoinContextTransformation, ProcessCompilationError, ValidationContext}
+import pl.touk.nussknacker.engine.api.typed.typing
 import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedObjectTypingResult}
 import pl.touk.nussknacker.engine.flink.api.process.{AbstractLazyParameterInterpreterFunction, FlinkCustomJoinTransformation, FlinkCustomNodeContext}
 import pl.touk.nussknacker.engine.flink.api.timestampwatermark.TimestampWatermarkHandler
 import pl.touk.nussknacker.engine.flink.util.timestamp.TimestampAssignmentHelper
 
 case object UnionTransformer extends UnionTransformer(None) {
 
-  val KeyField = "key"
-
   def transformContextsDefinition(valueByBranchId: Map[String, LazyParameter[AnyRef]], variableName: String)
                                  (inputContexts: Map[String, ValidationContext])
                                  (implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, ValidationContext] = {
+    val branchReturnTypes: Iterable[typing.TypingResult] = valueByBranchId.values.map(_.returnType)
     ContextTransformation.findUniqueParentContext(inputContexts).map { parent =>
-      val newType = TypedObjectTypingResult(
-        (UnionTransformer.KeyField -> Typed[String]) :: inputContexts.map {
-          case (branchId, _) =>
-            ContextTransformation.sanitizeBranchName(branchId) -> valueByBranchId(branchId).returnType
-        }.toList
-      )
-      ValidationContext(Map(variableName -> newType), Map.empty, parent)
+      ValidationContext(Map(variableName -> branchReturnTypes.head), Map.empty, parent)
+    }.andThen { vc =>
+      Validated.cond(branchReturnTypes.toSet.size == 1, vc, CannotCreateObjectError("All branch values must be of the same type", nodeId.id)).toValidatedNel
```

Comment:
Can we have some validation test, checking if this works ok? I also wonder - what about cases like `.canBeSubclassOf`? Is equality always enough?

Reply:
I've added test